renku/ui/api/models/plan.py

```diff
@@ -0,0 +1,213 @@
+in a Project. You can get a list of all active plans in a project by calling
+the ``list`` static method of each of these classes:
+
+.. code-block:: python
+
+    from renku.ui.api import Plan
+
+    plans = Plan.list()
+
+"""
+
+from datetime import datetime
+from typing import List, Optional, Union
+
+from renku.command.command_builder.database_dispatcher import DatabaseDispatcher
+from renku.core import errors
+from renku.domain_model.workflow import composite_plan as core_composite_plan
+from renku.domain_model.workflow import plan as core_plan
+from renku.infrastructure.gateway.plan_gateway import PlanGateway
+from renku.ui.api.models.project import ensure_project_context
+from renku.ui.api.models.run import Input, Link, Mapping, Output, Parameter
+
+
+class Plan:
+    """API Plan."""
+
+    def __init__(
+        self,
+        command: str,
+        date_created: Optional[datetime] = None,
+        description: Optional[str] = None,
+        inputs: List[Input] = None,
+        invalidated_at: Optional[datetime] = None,
+        keywords: Optional[List[str]] = None,
+        name: Optional[str] = None,
+        outputs: List[Output] = None,
+        parameters: List[Parameter] = None,
+        success_codes: Optional[List[int]] = None,
+    ):
+        self.command: str = command
+        self.date_created: Optional[datetime] = date_created
+        self.description: Optional[str] = description
+        self.inputs: List[Input] = inputs or []
+        self.invalidated_at: Optional[datetime] = invalidated_at
+        self.keywords: List[str] = keywords or []
+        self.name: Optional[str] = name
+        self.outputs: List[Output] = outputs or []
+        self.parameters: List[Parameter] = parameters or []
+        self.success_codes: List[int] = success_codes or []
+
+    @classmethod
+    def from_plan(cls, plan: core_plan.Plan) -> "Plan":
+        """Create an instance from a core Plan model.
+
+        Args:
+            plan(core_plan.Plan): The core plan.
+
+        Returns:
+            Plan: An API Plan model.
+        """
+        return cls(
+            command=plan.command,
+            date_created=plan.date_created,
+            description=plan.description,
+            inputs=[Input.from_parameter(i) for i in plan.inputs],
+            invalidated_at=plan.invalidated_at,
+            keywords=plan.keywords,
+            name=plan.name,
+            outputs=[Output.from_parameter(o) for o in plan.outputs],
+            parameters=[Parameter.from_parameter(p) for p in plan.parameters],
+            success_codes=plan.success_codes,
+        )
+
+    @staticmethod
+    def list(include_deleted: bool = False) -> List[Union["Plan", "CompositePlan"]]:
+        """List all plans in a project.
+
+        Args:
+            include_deleted(bool): Whether to include deleted plans.
+
+        Returns:
+            A list of all plans in the supplied project.
+        """
+        return _list_plans(include_deleted=include_deleted)
```

Comment:
Should we filter so `Plan.list` only returns `Plan` and `CompositePlan.list` only returns `CompositePlan` ?

Reply:
This is probably cleaner since CompositePlan doesn't inherit from Plan.